File: PyzeWatchOSInAppObjC/PyzeWatchOSInAppObjC WatchKit Extension/InAppMessageManager.m

#import "InAppMessageManager.h"


#pragma mark - Class InAppMessageManager

static InAppMessageManager * sharedInstance;

@interface InAppMessageManager ()

@property (strong, nonatomic) NSMutableArray *headers;
@property (assign, nonatomic) BOOL isRequested;

@end


@implementation InAppMessageManager

+ (instancetype) sharedManager {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[InAppMessageManager alloc] init];
    });
    return sharedInstance;
}

#pragma mark - Setters

-(void)setMessageActiveIndex:(NSInteger)messageActiveIndex {
    _messageActiveIndex = (messageActiveIndex < self.headers.count && messageActiveIndex > 0) ? messageActiveIndex : _messageActiveIndex;
}

#pragma mark - Public methods

- (void) countNewUnFetched:(void (^)(NSInteger))completionHandler {
    [Pyze countNewUnFetched:^(NSInteger count) {
        completionHandler(count);
    }];
}

- (void) updateIndex:(BOOL)shouldIncrement {
    self.messageActiveIndex = (shouldIncrement == NO) ? self.messageActiveIndex - 1 : self.messageActiveIndex + 1;
}

- (BOOL) isLastMessage {
    BOOL isLast = (self.messageActiveIndex == (self.inAppMessageCount - 1)) ?  YES : NO;
    return  isLast;
}

- (void) nextMessage:(void (^)(InAppModel*))completionHandler {
    [self updateIndex:YES];
    [self fetchMessage:^(InAppModel *messageObj) {
        completionHandler(messageObj);
    }];
}

- (void) previousMessage:(void (^)(InAppModel*))completionHandler {
    [self updateIndex:NO];
    [self fetchMessage:^(InAppModel *messageObj) {
        completionHandler(messageObj);
    }];
}

- (void) fetchHeadersForType:(PyzeInAppMessageType)type completionHandler:(void (^)(void))handler {
    
    /* Resetting content */
    self.messageActiveIndex = -1;
    self.activeInAppMessageType = type;
    [self.headers removeAllObjects];
    /* Resetting content */
    
    
    if (self.isRequested == NO) {
        
        self.isRequested = YES;
        
        [Pyze getMessageHeadersForType:type withCompletionHandler:^(NSArray *messageHeaders) {
            if (messageHeaders) {
                for (NSDictionary *header in messageHeaders) {
                    [self.headers addObject:header];
                }
                dispatch_async(dispatch_get_main_queue(), ^{
                    handler();
                });
                
            }
        }];
        
        
    }
    
}

#pragma mark - Private methods

- (void) fetchMessage:(void (^)(InAppModel * messageObj))completionHandler {
    NSDictionary *headerDict = [self messageHeader];
    [Pyze getMessageBodyWithCampaignID:headerDict[@"cid"]
                          andMessageID:headerDict[@"mid"]
                 withCompletionHandler:^(NSDictionary *messageBody) {
                     
        if (messageBody) {
            NSDictionary *payload = messageBody[@"payload"];
            if (payload) {
                InAppModel *inAppObj = [[InAppModel alloc] initWithDictionary:payload loadDefaults:NO];
                completionHandler(inAppObj);
            }
        }
    }];
}

- (NSDictionary *) messageHeader{
    return self.headers[self.messageActiveIndex];
}

@end


#pragma mark - Class InAppModel

@implementation InAppModel


-(instancetype) initWithDictionary : (NSDictionary *) inDictionary loadDefaults:(BOOL) load{
    
    self = [super init];
    if (self) {
        if (inDictionary && inDictionary.count) {
            self.msg = inDictionary[@"msg"];
            self.title = inDictionary[@"title"];
            self.templateID = inDictionary[@"templateId"];
            BOOL result = NO;
            if ([self.templateID isEqualToString:@"p_temp_1"])
                result = YES;
            if (inDictionary[@"templateData"]) {
                NSData * data = [inDictionary[@"templateData"] dataUsingEncoding:NSUTF8StringEncoding];
                NSDictionary * dicionary = [NSJSONSerialization JSONObjectWithData:data
                                                                           options:0
                                                                             error:nil];
                self.templateData = [[InAppTemplateData alloc] initWithDictionary:dicionary loadDefaults:load];
                self.templateData.textOverlapsImage = result;
            }
        }
    }
    return self;
}


@end



#pragma mark - Class InAppTemplateData

@implementation InAppTemplateData

-(instancetype) initWithDictionary : (NSDictionary *) inDictionary loadDefaults:(BOOL) load{
    
    self = [super init];
    if (self) {
        if (inDictionary && inDictionary.count)
        {
            self.imageURL = inDictionary[@"imageSourceURL"];
            self.optionalBase64Image = inDictionary[@"optionalBase64Image"];
            self.titleColorHex = inDictionary[@"titleColor"];
            self.msgColor = inDictionary[@"msgColor"];
            self.imageWidth = [inDictionary[@"imageWidth"] integerValue];
            self.imageHeight = [inDictionary[@"imageHeight"] integerValue];
            self.buttons = [NSArray arrayWithArray:inDictionary[@"buttons"]];
            self.canvasBgColor = inDictionary[@"canvas_bgColor"];
        }
    }
    return self;
}

@end

File: PyzeWatchOSInApp/PyzeWatchOSInApp WatchKit Extension/InAppMessageManager.swift
import Foundation
import Pyze

public class InAppMessageManager {
    
    private var headers:[Any] = []
    private var isRequested:Bool = false

    
    private var messageActiveIndex:Int {
        get {
            return self.messageActiveIndex
        }
        
        set {
            self.messageActiveIndex = (newValue < self.headers.count && newValue > 0) ? newValue : messageActiveIndex
        }
    }
    
    
    
    public var activeInAppMessageType:PyzeInAppMessageType {
        set {
            self.activeInAppMessageType = newValue
        }
        get {
            return self.activeInAppMessageType
        }
    }
    
    public var inAppMessageCount: Int = 0
    
    // MARK: - Init
    
    private init() {
        
    }
    
    
    // MARK: - Shared manager
    
    public static let sharedManager : InAppMessageManager = {
        let sharedManagerInstance = InAppMessageManager()
        sharedManagerInstance.messageActiveIndex = -1;
        return sharedManagerInstance
    } ()
    
    // MARK: - Public methods
    
    public func countNewUnFetched(completionhandler: @escaping (Int)->Void) -> Void {
        Pyze.countNewUnFetched { (count) in
            completionhandler(count)
        }
    }
    
    public func updateIndex(shouldIncrement:Bool) {
        self.messageActiveIndex = (shouldIncrement == false) ? self.messageActiveIndex - 1 : self.messageActiveIndex + 1
    }
    
    public func isLastMessage() -> Bool {
        let isLast = (self.messageActiveIndex == (self.inAppMessageCount - 1)) ? true : false
        return isLast
    }
    
    public func nextMessage (completionhandler:  @escaping (InAppModel)->Void) {
        self.updateIndex(shouldIncrement: true)
        self.fetchMessage { (inAppModelObj) in
            completionhandler(inAppModelObj)
        }
    }
    
    public func previousMessage (completionhandler:  @escaping (InAppModel)->Void) {
        self.updateIndex(shouldIncrement: false)
        self.fetchMessage { (inAppModelObj) in
            completionhandler(inAppModelObj)
        }
    }
    
    public func fetchHeaders(type:PyzeInAppMessageType, completionhandler: @escaping ()->Void) {
        
        /* Resetting content */
        self.messageActiveIndex = -1
        self.activeInAppMessageType = type
        self.headers.removeAll()
        /* Resetting content */
        
        if self.isRequested == false {
            
            self.isRequested = true
            Pyze.getMessageHeaders(for: type) { (messageHeadres) in
                
                self.isRequested = false
                if let mHeaders = messageHeadres {
                    self.inAppMessageCount = mHeaders.count
                    for headerDict in mHeaders {
                        self.headers.append(headerDict)
                        completionhandler()
                    }
                }
            }
        }
    }
    
    // MARK: - Private methods
    
    private func messageHeader() -> [String:Any]? {
        let header = self.headers[messageActiveIndex] as? [String:Any]
        return header ?? nil
    }
    
    private func fetchMessage(completionhandler:  @escaping (InAppModel)->Void) {
        
        if let dict = self.messageHeader() {
            Pyze.getMessageBody(withCampaignID: (dict["cid"] as? String ?? ""), andMessageID: (dict["mid"] as? String ?? "")) { (messageBody) in
                
                if let messageBody = messageBody {
                    
                    if let payload : [String:Any] = messageBody["payload"] as? [String:Any] {
                        let inAppObj:InAppModel = InAppModel.init(dictionary: payload, loadDefaults: false)
                        completionhandler(inAppObj)
                    }
                }
            }
        }
    }
}

// MARK: - Class InAppTemplateData

public class InAppTemplateData {
    
    var imageURL: String = ""
    var imageWidth: Int = 0
    var imageHeight: Int = 0
    var optionalBase64Image: String = ""
    var titleColorHex: String = ""
    var msgColor: String = ""
    var buttons = [Any]()
    var templateID: String = ""
    var canvasBgColor: String = ""
    var isTextOverlapsImage: Bool = false
    
    
    init(dictionary inDictionary: [String : Any], loadDefaults load: Bool) {
        
        self.imageURL = inDictionary["imageSourceURL"] as? String ?? ""
        self.optionalBase64Image = inDictionary["optionalBase64Image"] as? String ?? ""
        self.titleColorHex = inDictionary["titleColor"] as? String ?? ""
        self.msgColor = inDictionary["msgColor"] as? String ?? ""
        self.imageWidth = Int(inDictionary["imageWidth"] as? String ?? "") ?? 0
        self.imageHeight = Int(inDictionary["imageHeight"] as? String ?? "") ?? 0
        self.buttons = inDictionary["buttons"] as? [Any] ?? []
        self.canvasBgColor = inDictionary["canvas_bgColor"] as? String ?? ""
    }
}

// MARK: - Class InAppModel

public class InAppModel: NSObject {
    
    var msg = ""
    var title = ""
    var templateID = ""
    var templateData: InAppTemplateData!
    
    init(dictionary inDictionary: [String : Any], loadDefaults load: Bool) {
        
        self.msg = inDictionary["msg"] as? String ?? ""
        self.title = inDictionary["title"] as? String ?? ""
        self.templateID = inDictionary["templateId"] as? String ?? ""
        var result = false
        if (self.templateID == "p_temp_1") {
            result = true
        }
        do {
            if let templateData: String = inDictionary["templateData"] as? String {
                if let d:Data = templateData.data(using: .utf8) {
                    do {
                        let template: [String:Any]? = try JSONSerialization.jsonObject(with: d, options: JSONSerialization.ReadingOptions.mutableContainers) as? [String:Any]
                        self.templateData = InAppTemplateData(dictionary: template ?? [:], loadDefaults: load)
                        self.templateData.isTextOverlapsImage = result
                    } catch {
                        print("Error in parsing")
                    }
                }
            }
        }
    }
}



